### PR TITLE
Throttle render assets

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

allow throttling of gpu uploads to prevent choppy framerate when many textures/meshes are loaded in.

## Solution

- `RenderAsset`s can implement `byte_len()` which reports their size. implemented this for `Mesh` and `Image`
- users can add a `RenderAssetBytesPerFrame` which specifies max bytes to attempt to upload in a frame
- `render_assets::<A>` checks how many bytes have been written before attempting to upload assets. the limit is a soft cap: assets will be written until the total has exceeded the cap, to ensure some forward progress every frame

notes:
- this is a stopgap until we have multiple wgpu queues for proper streaming of data
- requires #12606

issues
- ~~fonts sometimes only partially upload. i have no clue why, needs to be fixed~~ fixed now.
- choosing the #bytes is tricky as it should be hardware / framerate dependent
- many features are not tested (env maps, light probes, etc) - they won't break unless `RenderAssetBytesPerFrame` is explicitly used though

### CODE REVIEW

**Comment by JMS55** [comment]
Is this about saving CPU time, or GPU/PCIE transfers?

For the former, we should move render asset preparation to background tasks (I wish we had a better way to mark systems as background instead of having to manually spawn tasks from a system). I have an issue open for this.

For the latter, yeah I think we need separate hardware queues. But I can see the value of this until then.

**Comment by robtfm** [comment]
it's about transfers. currently if you load a bunch of large renderassets the whole app stalls in `prepare_asset::<A>` until they're all done

**Comment by robtfm** [comment]
actually now you say that i'm not sure... `prepare_asset<Image>` uses `create_texture_with_data` which should just be submitting writes to the queue, so i don't know exactly why i see the stall point in `prepare_asset<Image>` itself. maybe it would be resolved using cpu-background uploading.

**Comment by IceSentry** [comment]
I just tried upgrading this to main. It's pretty simple, just need to do the same trick used for asset_usage where the param is `Self::SourceAsset` instead of `self`.

I tested it in a simple app that just spawns a bunch of helmets to see how it would behave and it doesn't seem to do anything. It still spawns every helmet at the same time with a major stutter. I never see the exhausted debug log. Maybe I broke something while porting it

File: crates/bevy_render/src/render_asset.rs [hunk 1]
```diff
@@ -38,6 +39,12 @@ pub trait RenderAsset: Asset + Clone {
     /// Whether or not to unload the asset after extracting it to the render world.
     fn asset_usage(&self) -> RenderAssetUsages;
 
+    /// Size of the data the asset will upload to the gpu. Specifying a return value
+    /// will allow the asset to be throttled via [`RenderAssetBytesPerFrame`].
+    fn byte_len(&self) -> Option<usize> {
```
Comment (IceSentry) [inline crates/bevy_render/src/render_asset.rs:44]:
```suggestion
    fn byte_len(source_asset: &Self::SourceAsset) -> Option<usize> {
```

This should make it compatible with main, just need to update the implementations.

**Comment by robtfm** [comment]
thanks for updating it - just making sure, you did add the resource as well? something like `app.insert_resource(RenderAssetBytesPerFrame::new(16777216))` ?

i definitely see the impact on 0.13 and can't see any reason it should have changed with the flip to implementing RenderAsset on targets ... although in my scenario images are by far the largest part of the problem, possibly i messed up the mesh size calculation.

**Comment by IceSentry** [comment]
Yeah, I did insert the resource, but I did it in a separate branch where I copied code manually so I may have missed something

File: crates/bevy_render/src/render_asset.rs [hunk 2]
```diff
@@ -322,14 +346,75 @@ pub fn prepare_assets<A: RenderAsset>(
         render_assets.remove(removed);
     }
 
-    for (id, extracted_asset) in extracted_assets.extracted.drain(..) {
+    let mut extracted_assets = extracted_assets.extracted.drain(..);
+
+    for (id, extracted_asset) in extracted_assets.by_ref() {
+        if let Some(size) = extracted_asset.byte_len() {
```
Comment (IceSentry) [inline crates/bevy_render/src/render_asset.rs:352]:
```suggestion
        if let Some(size) = A::byte_len(&extracted_asset) {
```

File: crates/bevy_render/src/texture/image.rs [hunk 3]
```diff
@@ -834,6 +834,10 @@ impl RenderAsset for Image {
         SRes<DefaultImageSampler>,
     );
 
+    fn byte_len(&self) -> Option<usize> {
```
Comment (IceSentry) [inline crates/bevy_render/src/texture/image.rs:837]:
```suggestion
    fn byte_len(image: &Self::SourceAsset) -> Option<usize> {
```

File: crates/bevy_render/src/mesh/mesh/mod.rs [hunk 4]
```diff
@@ -1430,6 +1430,18 @@ impl RenderAsset for Mesh {
...
+        let mut vertex_size = 0;
+        for attribute_data in self.attributes.values() {
+            let vertex_format = attribute_data.attribute.format;
+            vertex_size += vertex_format.get_size() as usize;
+        }
+
+        let vertex_count = self.count_vertices();
+        let index_bytes = self.get_index_buffer_bytes().map(<[_]>::len).unwrap_or(0);
+        Some(vertex_size * vertex_count + index_bytes)
+    }
```
Comment (IceSentry) [inline crates/bevy_render/src/mesh/mesh/mod.rs:1443]:
```suggestion
    fn byte_len(mesh: &Self::SourceAsset) -> Option<usize> {
        let mut vertex_size = 0;
        for attribute_data in mesh.attributes.values() {
            let vertex_format = attribute_data.attribute.format;
            vertex_size += vertex_format.get_size() as usize;
        }

        let vertex_count = mesh.count_vertices();
        let index_bytes = mesh.get_index_buffer_bytes().map(<[_]>::len).unwrap_or(0);
        Some(vertex_size * vertex_count + index_bytes)
    }
```

File: crates/bevy_render/src/texture/image.rs [hunk 5]
```diff
@@ -834,6 +834,10 @@ impl RenderAsset for Image {
...
+        Some(self.data.len())
+    }
```
Comment (IceSentry) [inline crates/bevy_render/src/texture/image.rs:839]:
```suggestion
    fn byte_len(image: &Self::SourceAsset) -> Option<usize> {
    	Some(image.data.len())
    }
```

**Comment by IceSentry** [comment]
I tested again based on the branch before the changes and it did work, so I probably messed up something when I was copying the code.

**Comment by robtfm** [comment]
ok cool, i'll go through properly at some point this weekend.

**Comment by robtfm** [comment]
i got a bit tangled up with the merge so i just applied the changes onto a new branch and merged it in (with a couple of tweaks).

i can't really test very well since my project is not on main - i tried with 20 copies of the flight helmet and it seems to throttle correctly according to the debug output, but its not really enough data to see the impact visually.

**Comment by mockersf** [comment]
tried on a project I have that loads 65536 meshes with their textures, this works as expected. Bevy freezes without setting `RenderAssetBytesPerFrame`, works fine with it. FPS is still bad because of the number unique meshes / textures, and you're not supposed to display all the high details one at the same time 😄 

https://github.com/bevyengine/bevy/assets/8672791/9f0b47a1-3fe6-4e76-9e19-0d0fab4f1a82

File: crates/bevy_render/src/render_asset.rs [hunk 6]
```diff
@@ -322,13 +356,91 @@ pub fn prepare_assets<A: RenderAsset>(
...
+                bpf.write_bytes(write_bytes);
+                wrote_asset_count += 1;
             }
             Err(PrepareAssetError::RetryNextUpdate(extracted_asset)) => {
                 prepare_next_frame.assets.push((id, extracted_asset));
             }
         }
     }
+
+    if bpf.exhausted() {
```
Comment (mockersf) [inline crates/bevy_render/src/render_asset.rs:386]: this is logging too much, as soon as one asset exhausted the budget, all the others will log even if they don't have anything to do

```suggestion
    if bpf.exhausted() && !prepare_next_frame.assets.is_empty() {
```

**Review by mockersf — APPROVED** [review]

**Comment by JMS55** [comment]
@mockersf can we add this as a cargo example? I may work on further asset loading perf improvements in the future, and it would be amazing to have this as a test bench.

**Review by IceSentry — APPROVED** [review]
LGTM

Would be nice to have a large_asset_loading example or something like that, but I don't want to block this PR on that.

**Comment by mockersf** [comment]
> @mockersf can we add this as a cargo example? I may work on further asset loading perf improvements in the future, and it would be amazing to have this as a test bench.

I can do that in a few days, but it'll definitely never be merged in Bevy given how ugly it is 😄 
Everything is loaded from a 65MB 7zip file

> Would be nice to have a large_asset_loading example or something like that, but I don't want to block this PR on that.

I can try to get something prettier (and smaller) than what I currently have later

File: crates/bevy_render/src/render_asset.rs [hunk 7]
```diff
@@ -299,17 +314,36 @@ pub fn prepare_assets<A: RenderAsset>(
...
     param: StaticSystemParam<<A as RenderAsset>::Param>,
+    mut bpf: ResMut<RenderAssetBytesPerFrame>,
 ) {
+    let mut wrote_asset_count = 0;
+
     let mut param = param.into_inner();
     let queued_assets = std::mem::take(&mut prepare_next_frame.assets);
     for (id, extracted_asset) in queued_assets {
-        if extracted_assets.removed.contains(&id) {
+        if extracted_assets.removed.contains(&id) || extracted_assets.added.contains(&id) {
```
Comment (JMS55) [inline crates/bevy_render/src/render_asset.rs:324]: What is the purpose of skipping newly added assets here?

File: crates/bevy_render/src/render_asset.rs [hunk 8]
```diff
@@ -161,13 +170,15 @@ impl<A: RenderAsset> RenderAssetDependency for A {
 pub struct ExtractedAssets<A: RenderAsset> {
     extracted: Vec<(AssetId<A::SourceAsset>, A::SourceAsset)>,
     removed: Vec<AssetId<A::SourceAsset>>,
+    added: Vec<AssetId<A::SourceAsset>>,
```
Comment (JMS55) [inline crates/bevy_render/src/render_asset.rs:173]: Should be a HashSet over a Vec.

**Review by JMS55 — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/render_asset.rs (hunk 7 above)
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs:324]: This is the retry loop for assets that were previously extracted and were not ready in the previous frame. If the asset has been added again, there are now 2 versions and we don’t want to prepare the old one. 

This was the source of the font bug I had where some characters would not render:

The code in prepare_sprites clears its cache on asset change events, then checks for presence in RenderAssets to repopulate the cache. If we prepare the old version then the code in the sprite module will see it as ready and will cache the previous version, resulting in the old asset being permanently bound for some glyphs.

File: crates/bevy_render/src/render_asset.rs (hunk 8 above)
Comment (robtfm) [reply crates/bevy_render/src/render_asset.rs:173]: Good point, will change this and `removed` as well

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1306 if you'd like to help out.

**Pull request merged**